File: cpp/tests/test_cross_binding_integration.cpp
// Cross-binding integration test (R18 cluster 5 — Cat 33d).
//
// Counterpart of python/tests/test_cross_binding_integration.py and
// go/aletheia/cross_binding_integration_test.go. All three tests construct
// identical canonical inputs in code (no shared corpus, no golden output to
// diff against) and assert the binding's response shape matches the
// structural invariants documented in docs/architecture/PROTOCOL.md.
//
// The shared truth is PROTOCOL.md (text + JSON examples). Corpus-style
// integration tests bit-rot fast and tie every binding to one binding's exact
// emit format. By asserting structural invariants here (member presence,
// value types, count relationships, error-code identity), each binding's
// drift from the documented protocol surfaces locally without depending on
// the other bindings being run.

#include <catch2/catch_test_macros.hpp>

#include <aletheia/aletheia.hpp>

#include <array>
#include <cstdlib>
#include <filesystem>
#include <span>
#include <stop_token>
#include <variant>
#include <vector>

using namespace aletheia;
namespace fs = std::filesystem;

namespace {

// Matches python/tests/test_cross_binding_integration.py::_CANONICAL_DBC and
// go/aletheia/cross_binding_integration_test.go::canonicalDBC. Drift of any
// of the three is the cross-binding hazard the test is designed to catch.
auto canonical_dbc() -> DbcDefinition {
    auto sig_id = StandardId::create(256).value();
    auto dlc = Dlc::create(8).value();
    DbcSignal test_sig{
        .name = SignalName{"TestSignal"},
        .start_bit = BitPosition{0},
        .bit_length = BitLength{16},
        .byte_order = ByteOrder::LittleEndian,
        .is_signed = false,
        .factor = RationalFactor{Rational{1, 1}},
        .offset = RationalOffset{Rational{0, 1}},
        .minimum = RationalBound{Rational{0, 1}},
        .maximum = RationalBound{Rational{65535, 1}},
        .unit = Unit{""},
        .presence = AlwaysPresent{},
    };
    return DbcDefinition{
        .version = "1.0",
        .messages = {DbcMessage{
            .id = CanId{sig_id},
            .name = MessageName{"TestMessage"},
            .dlc = dlc,
            .sender = NodeName{"ECU"},
            .signals = {test_sig},
        }},
    };
}

auto find_lib() -> fs::path {
    if (auto* env = std::getenv("ALETHEIA_LIB"))
        return env;
    auto project_root = fs::path{__FILE__}.parent_path().parent_path().parent_path();
    auto lib = project_root / "build" / "libaletheia-ffi.so";
    if (fs::exists(lib))
        return lib;
    SKIP("libaletheia-ffi.so not found — run 'cabal run shake -- build' first");
    return {};
}

} // namespace

TEST_CASE("ParsedDBC response has documented shape", "[cross_binding]") {
    auto lib = find_lib();
    auto backend = make_ffi_backend(lib);
    AletheiaClient client(std::move(backend));

    auto result = client.parse_dbc(std::stop_token{}, canonical_dbc());
    REQUIRE(result.has_value());
    const auto& parsed = result.value();

    // ParsedDBC: { dbc: DbcDefinition, warnings: vector<ValidationIssue> }.
    // Round-trip identity invariant on canonical content.
    REQUIRE(parsed.dbc.messages.size() == 1);
    CHECK(parsed.dbc.messages[0].name.get() == "TestMessage");
    REQUIRE(parsed.dbc.messages[0].signals.size() == 1);
    CHECK(parsed.dbc.messages[0].signals[0].name.get() == "TestSignal");
    // Warnings vector must be empty (canonical DBC is well-formed); the
    // member's presence + size-zero is the cross-binding parity check.
    CHECK(parsed.warnings.empty());
}

TEST_CASE("ValidationResult response has documented shape", "[cross_binding]") {
    auto lib = find_lib();
    auto backend = make_ffi_backend(lib);
    AletheiaClient client(std::move(backend));

    auto result = client.validate_dbc(std::stop_token{}, canonical_dbc());
    REQUIRE(result.has_value());
    const auto& validation = result.value();

    // ValidationResult: { has_errors: bool, issues: vector<ValidationIssue> }.
    CHECK_FALSE(validation.has_errors);
    CHECK(validation.issues.empty());
}

TEST_CASE("send_frame ack response has documented shape", "[cross_binding]") {
    auto lib = find_lib();
    auto backend = make_ffi_backend(lib);
    AletheiaClient client(std::move(backend));

    REQUIRE(client.parse_dbc(std::stop_token{}, canonical_dbc()).has_value());
    auto formula = ltl::always(
        ltl::atomic(ltl::less_than(SignalName{"TestSignal"}, PhysicalValue{Rational{1000, 1}})));
    std::vector<LtlFormula> props;
    props.push_back(std::move(formula));
    REQUIRE(client.set_properties(std::stop_token{}, props).has_value());
    REQUIRE(client.start_stream(std::stop_token{}).has_value());

    auto sid = StandardId::create(256).value();
    auto dlc = Dlc::create(8).value();
    auto payload = std::array<std::byte, 8>{};
    auto resp = client.send_frame(std::stop_token{}, Timestamp{1000}, CanId{sid}, dlc,
                                  std::span<const std::byte>{payload});
    REQUIRE(resp.has_value());
    // FrameResponse = variant<Ack, Violation>; signal value 0 < 1000 → Ack.
    CHECK(std::holds_alternative<Ack>(resp.value()));
}

TEST_CASE("send_frame violation response has documented shape", "[cross_binding]") {
    auto lib = find_lib();
    auto backend = make_ffi_backend(lib);
    AletheiaClient client(std::move(backend));

    REQUIRE(client.parse_dbc(std::stop_token{}, canonical_dbc()).has_value());
    auto formula = ltl::always(
        ltl::atomic(ltl::less_than(SignalName{"TestSignal"}, PhysicalValue{Rational{100, 1}})));
    std::vector<LtlFormula> props;
    props.push_back(std::move(formula));
    REQUIRE(client.set_properties(std::stop_token{}, props).has_value());
    REQUIRE(client.start_stream(std::stop_token{}).has_value());

    auto sid = StandardId::create(256).value();
    auto dlc = Dlc::create(8).value();
    // Signal value 0xFFFF (65535) > 100 → violation.
    auto payload = std::array<std::byte, 8>{
        std::byte{0xFF}, std::byte{0xFF}, std::byte{0}, std::byte{0},
        std::byte{0},    std::byte{0},    std::byte{0}, std::byte{0},
    };
    auto resp = client.send_frame(std::stop_token{}, Timestamp{1000}, CanId{sid}, dlc,
                                  std::span<const std::byte>{payload});
    REQUIRE(resp.has_value());
    REQUIRE(std::holds_alternative<Violation>(resp.value()));
    const auto& v = std::get<Violation>(resp.value());
    // Documented Violation members: property_index, timestamp, reason,
    // optional<enrichment>; assert the timestamp survived round-trip.
    CHECK(v.timestamp.count() > 0);
}

TEST_CASE("invalid CAN ID is rejected at type boundary", "[cross_binding]") {
    // CAN ID 0x800 = 2048 is out of standard 11-bit range; StandardId::create
    // rejects it before the FFI is reached. This is the C++ analogue of
    // Python raising on out-of-range and Go returning an Error variant — the
    // cross-binding invariant is "invalid CAN ID is rejected somewhere on
    // the path", satisfied here by the strong-typed factory.
    auto sid = StandardId::create(0x800);
    CHECK_FALSE(sid.has_value());

    // Extended ID just over the 29-bit cap (2^29 = 0x20000000) is also rejected.
    auto xid = ExtendedId::create(0x20000000);
    CHECK_FALSE(xid.has_value());
}

// R19 cluster 8 phase e.1 — Identifier validity record enforces
// max_identifier_length.  The Agda kernel's `validIdentifierᵇ` predicate
// gained a third conjunct asserting `length name <ᵇ suc max-identifier-
// length`.  Identifiers at the limit (128 chars) still parse; anything
// longer is rejected at `mkIdentFromChars` and surfaces as a parse error
// on the wire (currently `dbc_text_trailing_input` due to parser-monad
// position semantics; refining to typed `InputBoundExceeded
// IdentifierLength` is downstream parser-monad plumbing).

TEST_CASE("identifier at max length is accepted", "[cross_binding][cluster8][e1]") {
    auto lib = find_lib();
    auto backend = make_ffi_backend(lib);
    AletheiaClient client(std::move(backend));

    const std::string name(aletheia::max_identifier_length, 'A');
    const std::string dbc_text = "VERSION \"\"\nNS_:\nBS_:\nBU_:\nBO_ 100 " + name + ": 8 ECU\n";
    auto result = client.parse_dbc_text(std::stop_token{}, dbc_text);
    REQUIRE(result.has_value());
    REQUIRE(result->dbc.messages.size() == 1);
    CHECK(result->dbc.messages[0].name.get() == name);
}

TEST_CASE("identifier over max length is rejected", "[cross_binding][cluster8][e1]") {
    auto lib = find_lib();
    auto backend = make_ffi_backend(lib);
    AletheiaClient client(std::move(backend));

    const std::string name(aletheia::max_identifier_length + 1, 'A');
    const std::string dbc_text = "VERSION \"\"\nNS_:\nBS_:\nBU_:\nBO_ 100 " + name + ": 8 ECU\n";
    auto result = client.parse_dbc_text(std::stop_token{}, dbc_text);
    REQUIRE_FALSE(result.has_value());
}

// R19 AGDA-D-13.4 phase 2a — typed NestingDepth wire-error.  A
// deeply-nested LTL formula triggers the kernel's `jsonDepth` check at
// `handleParsedJSON`, emitting `ParseErr (InputBoundExceeded NestingDepth …)`
// instead of the pre-phase-2a untyped `DispatchErr InvalidJSON`.  The wire
// carries `bound_kind / observed / limit` which `make_json_error` lifts
// into `AletheiaError::bound_info()`.  Mirrors Python's
// `TestNestingDepthBound::test_nested_at_depth_63_rejected` and Go's
// `TestCrossBinding_NestingDepthLiftsToInputBoundExceeded`.
TEST_CASE("nesting depth over limit lifts to InputBoundExceeded",
          "[cross_binding][cluster8][phase2a]") {
    auto lib = find_lib();
    auto backend = make_ffi_backend(lib);
    AletheiaClient client(std::move(backend));

    REQUIRE(client.parse_dbc(std::stop_token{}, canonical_dbc()).has_value());
    // 63 always-wrappers + atomic + predicate = JSON depth 65 (> 64).
    LtlFormula inner =
        ltl::atomic(ltl::equals(SignalName{"TestSignal"}, PhysicalValue{Rational{0, 1}}));
    for (std::size_t i = 0; i < 63; ++i)
        inner = ltl::always(std::move(inner));
    std::vector<LtlFormula> props;
    props.push_back(std::move(inner));
    auto result = client.set_properties(std::stop_token{}, props);
    REQUIRE_FALSE(result.has_value());
    CHECK(result.error().kind() == ErrorKind::InputBoundExceeded);
    CHECK(result.error().code() == ErrorCode::ParseInputBoundExceeded);
    REQUIRE(result.error().bound_info().has_value());
    CHECK(result.error().bound_info()->bound_kind == bound_kind_nesting_depth);
    CHECK(result.error().bound_info()->limit == max_nesting_depth);
    CHECK(result.error().bound_info()->observed > max_nesting_depth);
}

File: python/tests/test_cross_binding_integration.py
"""Cross-binding integration tests.

Counterpart of ``go/aletheia/cross_binding_integration_test.go`` and
``cpp/tests/test_cross_binding_integration.cpp``.  All three tests
construct identical canonical inputs in code (no shared corpus, no
golden output to diff against) and assert the binding's response shape
matches the structural invariants documented in
``docs/architecture/PROTOCOL.md``.

The shared truth source is PROTOCOL.md (text + JSON examples), not a
file of recorded outputs — corpus-style integration tests bit-rot fast
and tie every binding to one binding's exact emit format.  By
asserting structural invariants here (key sets, value types, count
relationships, error-code enum membership), each binding's drift from
the documented protocol surfaces locally without depending on the
other bindings to also have run.

Cross-binding parity is enforced transitively: if Python's test asserts
"``ParsedDBCResponse`` has exactly the keys ``{status, dbc, issues}``"
and Go's test asserts the same, any binding that adds or drops a key
fails its own test.  No pairwise binding diff is computed.

Per AGENTS.md cat 22, Python is the reference implementation — but
"reference" means "the documented shape lives in Python's TypedDicts";
the test suite below validates every binding against PROTOCOL.md, not
against Python's runtime output.
"""

from __future__ import annotations

import pytest
from _canonical_dbc import CANONICAL_DBC as _CANONICAL_DBC

from aletheia import AletheiaClient, Signal, ValidationError
from aletheia.protocols import DLCCode

# Documented structural invariants — these mirror PROTOCOL.md's response
# shape tables.  Each binding asserts the same expectations against its
# own deserialized output.

_PARSED_DBC_RESPONSE_KEYS: frozenset[str] = frozenset({"status", "dbc", "warnings"})
_VALIDATION_RESPONSE_KEYS: frozenset[str] = frozenset({"status", "has_errors", "issues"})
_ACK_RESPONSE_KEYS: frozenset[str] = frozenset({"status"})
_ERROR_RESPONSE_KEYS: frozenset[str] = frozenset({"status", "code", "message"})


class TestCrossBindingIntegration:
    """Structural-invariant assertions across the public response surface.

    Mirrored verbatim by ``go/aletheia/cross_binding_integration_test.go``
    and ``cpp/tests/test_cross_binding_integration.cpp``.  When this
    file's expectations change, the other two MUST change in lock-step.
    """

    def test_parse_dbc_response_shape(self) -> None:
        """``parse_dbc`` returns a structure matching ``ParsedDBCResponse``."""
        with AletheiaClient() as client:
            response = client.parse_dbc(_CANONICAL_DBC)

        # ParsedDBCResponse: status="success", dbc=<DBCDefinition>, warnings=[].
        assert isinstance(response, dict)
        assert _PARSED_DBC_RESPONSE_KEYS.issubset(set(response.keys())), (
            f"parse_dbc response missing keys: {_PARSED_DBC_RESPONSE_KEYS - set(response.keys())}"
        )
        assert response["status"] == "success"
        assert isinstance(response["warnings"], list)
        # Round-trip identity invariant: parsed body retains canonical DBC's
        # structural content (one message, one signal, name preserved).
        parsed_dbc = response["dbc"]
        assert isinstance(parsed_dbc, dict)
        assert isinstance(parsed_dbc.get("messages"), list)
        assert len(parsed_dbc["messages"]) == 1
        assert parsed_dbc["messages"][0]["id"] == 256
        assert parsed_dbc["messages"][0]["name"] == "TestMessage"
        assert len(parsed_dbc["messages"][0]["signals"]) == 1
        assert parsed_dbc["messages"][0]["signals"][0]["name"] == "TestSignal"

    def test_validate_dbc_response_shape(self) -> None:
        """``validate_dbc`` returns a structure matching ``ValidationResponse``."""
        with AletheiaClient() as client:
            response = client.validate_dbc(_CANONICAL_DBC)

        # ValidationResponse: status="validation", has_errors=bool, issues=list.
        assert isinstance(response, dict)
        assert _VALIDATION_RESPONSE_KEYS.issubset(set(response.keys())), (
            f"validate_dbc response missing keys: "
            f"{_VALIDATION_RESPONSE_KEYS - set(response.keys())}"
        )
        assert response["status"] == "validation"
        assert isinstance(response["has_errors"], bool)
        assert response["has_errors"] is False  # canonical DBC is valid
        assert isinstance(response["issues"], list)

    def test_send_frame_ack_response_shape(self) -> None:
        """``send_frame`` on a non-violating frame returns ``AckResponse``."""
        prop = Signal("TestSignal").less_than(1000).always()
        with AletheiaClient() as client:
            client.parse_dbc(_CANONICAL_DBC)
            client.set_properties([prop.to_dict()])
            client.start_stream()
            response = client.send_frame(
                timestamp=1000,
                can_id=256,
                dlc=DLCCode(8),
                data=bytearray([0, 0, 0, 0, 0, 0, 0, 0]),  # signal value 0 < 1000
            )
            client.end_stream()

        # AckResponse: status="ack", no other required keys.
        assert isinstance(response, dict)
        assert "status" in response  # narrows the union away from PropertyBatchResponse
        assert response["status"] == "ack"
        assert _ACK_RESPONSE_KEYS.issubset(set(response.keys()))

    def test_send_frame_violation_response_shape(self) -> None:
        """``send_frame`` on a violating frame returns ``PropertyBatchResponse``.

        Required wire keys per PROTOCOL.md:
        ``{type: "property_batch", results: [...]}``; each ``results``
        entry has ``{type: "property", status: "fails"|"holds"|"unresolved",
        property_index, [timestamp]}``.  ``reason`` and ``enrichment`` are
        optional but conventionally present on fails entries when the
        binding has the DBC + property set loaded.
        """
        prop = Signal("TestSignal").less_than(100).always()
        with AletheiaClient() as client:
            client.parse_dbc(_CANONICAL_DBC)
            client.set_properties([prop.to_dict()])
            client.start_stream()
            # Signal value 0xFFFF (65535) > 100 → violation.
            response = client.send_frame(
                timestamp=1000,
                can_id=256,
                dlc=DLCCode(8),
                data=bytearray([0xFF, 0xFF, 0, 0, 0, 0, 0, 0]),
            )
            client.end_stream()

        assert isinstance(response, dict)
        assert "type" in response  # narrows the union to PropertyBatchResponse
        assert response["type"] == "property_batch"
        assert isinstance(response["results"], list)
        assert len(response["results"]) >= 1
        # Violations are always the last entry in source-order per the
        # Agda dispatchIterResult invariant (satisfactions first, halt last).
        violation = response["results"][-1]
        assert violation["status"] == "fails"
        assert violation["type"] == "property"
        assert "property_index" in violation
        assert "timestamp" in violation

    def test_send_frame_error_response_shape(self) -> None:
        """Invalid frame input returns ``ErrorResponse`` with required fields."""
        with AletheiaClient() as client:
            client.parse_dbc(_CANONICAL_DBC)
            client.start_stream()
            # CAN ID 0x800 (2048) is out of standard 11-bit range; FFI rejects.
            with pytest.raises((ValidationError, ValueError, RuntimeError, OSError)):
                client.send_frame(
                    timestamp=1000,
                    can_id=0x800,
                    dlc=DLCCode(8),
                    data=bytearray([0, 0, 0, 0, 0, 0, 0, 0]),
                )

        # Error path is exception-based in Python; the wire-form
        # ErrorResponse shape is enforced by the binding's send_frame
        # implementation translating wire errors to Python exceptions.
        # The cross-binding parity claim is that an invalid CAN ID
        # produces an error in EVERY binding (Go: error return; C++:
        # error variant; Python: raise) — not a silent ack.

    def test_error_response_keys_when_observed_directly(self) -> None:
        """When an ErrorResponse surfaces as a dict (not exception), it has the documented keys.

        Currently Python's high-level API converts errors to exceptions,
        so this test exercises the wire-format keyset assertion that
        Go and C++ enforce on the dict-form response.
        """
        # Synthesize a wire-form ErrorResponse and assert its key set —
        # this is the static structural invariant that the binding's
        # parser must accept.
        synthetic: dict[str, str] = {
            "status": "error",
            "code": "frame_invalid_can_id",
            "message": "CAN ID out of range",
        }
        assert _ERROR_RESPONSE_KEYS.issubset(set(synthetic.keys()))
        assert synthetic["status"] == "error"
